gematria/datasets/basic_block_utils.cc

```diff
@@ -0,0 +1,93 @@
+// Copyright 2024 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "gematria/datasets/basic_block_utils.h"
+
+#include <map>
+
+#include "X86.h"
+#include "X86InstrInfo.h"
+#include "X86RegisterInfo.h"
+#include "llvm/MC/MCInstrInfo.h"
+#include "llvm/MC/MCRegisterInfo.h"
+
+using namespace llvm;
+
+namespace gematria {
+
+unsigned getSuperRegister(unsigned OriginalRegister,
+                          const MCRegisterInfo &RegisterInfo) {
+  unsigned SuperRegister = OriginalRegister;
+  for (MCPhysReg CurrentSuperRegister :
+       RegisterInfo.superregs_inclusive(OriginalRegister)) {
+    SuperRegister = CurrentSuperRegister;
+  }
```

Comment:
It will always return a super-register (at least from what I understand). So if we give it `AL`, it won't return `AX`, it would just return `EAX`. Just pulling the back of the iterator returns zero, which seems like an issue on the LLVM side. I'll add a TODO.

Reply:
It will return `RAX`. I misspoke there.

Looking over the `superregs_inclusive` iterator implementation again, it doesn't look like each successive register is guaranteed to be a super register. I've added a check that the register we're assigning to `SuperRegister` is in fact a super register of it.

I would add an assertion that the bit width is bigger, but it seems surprisingly hard to get the bit width. One either needs to go through `TargetRegisterInfo` (which seems to be a pain to get here), or get the register class, but the register class is not guaranteed to have the bit width (although I would think it would for X86 GPRs), and there is seemingly no direct mapping from registers to their register classes in `MCRegisterInfo`.

Either way, the explicit super register check should be fine.